### PR TITLE
feat: use tailnet v2 API for coordination

### PR DESCRIPTION
**Pull request by spikecurtis**
This one is huge, and I'm sorry.

The problem is that once I change `tailnet.Conn` to start doing v2 behavior, I kind of have to change it everywhere, including in CoderSDK (CLI), the agent, wsproxy, and ServerTailnet.

There is still a bit more cleanup to do, and I need to add code so that when we lose connection to the Coordinator, we mark all peers as LOST, but that will be in a separate PR since this is big enough!

### CODE REVIEW

**Comment by spikecurtis** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/coder/coder/11638?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

<b>In this stack:</b> <code>🚀 Improve Tailnet Coordination and Functionality</code><a href="https://withgraphite.notion.site/Graphite-beta-features-EXTERNAL-496843566b204c7a97a2a21f8dea5434">(?)</a>

* `main`
  * **#11398** <a href="https://app.graphite.dev/github/pr/coder/coder/11398?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **#11399** <a href="https://app.graphite.dev/github/pr/coder/coder/11399?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **#11400** <a href="https://app.graphite.dev/github/pr/coder/coder/11400?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **#11487** <a href="https://app.graphite.dev/github/pr/coder/coder/11487?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **#11512** <a href="https://app.graphite.dev/github/pr/coder/coder/11512?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **#11539** <a href="https://app.graphite.dev/github/pr/coder/coder/11539?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
              * **#11567** <a href="https://app.graphite.dev/github/pr/coder/coder/11567?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                * **#11568** <a href="https://app.graphite.dev/github/pr/coder/coder/11568?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                  * **#11569** <a href="https://app.graphite.dev/github/pr/coder/coder/11569?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                    * **#11571** <a href="https://app.graphite.dev/github/pr/coder/coder/11571?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                      * **#11590** <a href="https://app.graphite.dev/github/pr/coder/coder/11590?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                        * **#11634** <a href="https://app.graphite.dev/github/pr/coder/coder/11634?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                          * **#11635** <a href="https://app.graphite.dev/github/pr/coder/coder/11635?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                            * **#11636** <a href="https://app.graphite.dev/github/pr/coder/coder/11636?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                              * **#11637** <a href="https://app.graphite.dev/github/pr/coder/coder/11637?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                                * **#11638** <a href="https://app.graphite.dev/github/pr/coder/coder/11638?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @spikecurtis and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by coadler — APPROVED** [review]
Looks good. Nice to see it all come together

File: agent/agenttest/client.go [hunk 1]
```diff
@@ -24,18 +31,39 @@ func NewClient(t testing.TB,
...
 		manifest.AgentID = agentID
 	}
+	coordPtr := atomic.Pointer[tailnet.Coordinator]{}
+	coordPtr.Store(&coordinator)
+	mux := drpcmux.New()
+	drpcService := &tailnet.DRPCService{
+		CoordPtr: &coordPtr,
+		Logger:   logger,
+		// TODO: handle DERPMap too!
+		DerpMapUpdateFrequency: time.Hour,
```
[the remark this answers is not in the document]
Comment (spikecurtis) [reply agent/agenttest/client.go:46]: Addressed in some upstack PRs.  As of this PR we still use the old DERPMap dedicated websocket, but I'm in the process of switching everything over after this one.

**Comment by spikecurtis** [comment]
### Merge activity

* **Jan 22, 2:07 AM**: @spikecurtis merged this pull request with [Graphite](https://app.graphite.dev/github/pr/coder/coder/11638).

**Pull request merged**